androidenv: upgrade to repository2-3

hadilq opened the pull request:
AndroidEnv is relying on XML files provided by Google(available in https://dl.google.com/android/repository/) so we need to update them if Google bumped their version up. Here we updated to repository2-3.xml.

Thanks to @adrian-gierakowski to figured this out in https://github.com/NixOS/nixpkgs/issues/359918!


## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

hadilq committed:
androidenv: upgradde to repository2-3

androidenv: apply numinit patch to have a nice error message

androidenv handle unsupported systems, and make tests running by default!

androidenv run nixfmt

androidenv: support emulator/systemimages/NDK only for certain systems

androidenv: apply nixfmt

androidenv: give a warning if archive doesn't exist

adrian-gierakowski commented on pkgs/development/mobile/androidenv/repo.json:
```diff
@@ -15331,15 +18252,91 @@
+            "size": 304191081,
+            "url": "https://dl.google.com/android/repository/emulator-linux_x64-12658423.zip"
+          },
+          {
+            "os": "macosx",
+            "sha1": "88d045da40b76e9ff0893c4b477866bf0119ecd1",
+            "size": 320263693,
+            "url": "https://dl.google.com/android/repository/emulator-darwin_aarch64-12658423.zip"
+          },
```
I believe this will break things for people on `x86_64-darwin` 

in addition to updating the repo urls, we also need to adapt the script which generates `repo.json` to make a distinction between `x86_64-darwin` and `darwin_aarch64` and finally adapt the nix code [here](https://github.com/NixOS/nixpkgs/blob/74831549826cb12b58952e92b6827e2e2110b25f/pkgs/development/mobile/androidenv/compose-android-packages.nix#L30-L31) (and everywhere where the `os` variable is used), to choose the correct architecture for a package.

hadilq replied:
Interesting! You are right! Let me check! We may need to iterate like this multiple times, since I don't have access to an aarch64 CPU, right now!

adrian-gierakowski replied:
Sure. I’m a Linux user myself but I can test this out on macOS, both aarch64 and x86, on github actions 

thanks!

hadilq replied:
@adrian-gierakowski please review and test! Thanks!

adrian-gierakowski replied:
we're getting problem, but I think we still have some issues:
1. we are missing x86_64-darwin versions of the emulator in [repo.json.](https://github.com/NixOS/nixpkgs/blob/a7927a90b7c16703a01189693639366ec4f31132/pkgs/development/mobile/androidenv/repo.json#L18857-L18879) (compare that to [android-nixpkgs](https://github.com/tadfisher/android-nixpkgs/blob/6fc652b651d3cd2c7d40f1f6a40630041cc0687a/channels/beta/default.nix#L3005-L3021) which has 4 entries).
2. seem like "arch" is missing for some versions of emulator (and maybe other packades) in `repo.json`, for example it's present [here](https://github.com/NixOS/nixpkgs/blob/a7927a90b7c16703a01189693639366ec4f31132/pkgs/development/mobile/androidenv/repo.json#L18737-L18760) but missing [here](https://github.com/NixOS/nixpkgs/blob/a7927a90b7c16703a01189693639366ec4f31132/pkgs/development/mobile/androidenv/repo.json#L18700-L18719)

consequence of point 1 is that we get correct binary on aarch64-darwin as can be seen in [this test run](https://github.com/rhinofi/android-nixpkgs-test/actions/runs/12096436144/job/33730732869)) 

![image](https://github.com/user-attachments/assets/86cf88f5-ba71-423b-aa6a-ab2d218deea9)


but get a build (or rather evaluation) failure x86_64-dawrwin (see [here](https://github.com/rhinofi/android-nixpkgs-test/actions/runs/12096436145/job/33730732850))
 
consequence of point 2 is that for some versions of emulator (35.1.4) we're get the old behaviour

hadilq replied:
@adrian-gierakowski Sorry for late reply! I pushed a commit to address these issues. Let me know if this one works :D

adrian-gierakowski reviewed (changes requested):
thanks for taking the time to look into this so quickly!

I believe further work is required though, see incline comment.

hadilq commented:
Hey @adrian-gierakowski , let's ping you here again. I pushed the fix almost two weeks ago. So let me know your thoughts.

adrian-gierakowski commented:
> Hey @adrian-gierakowski , let's ping you here again. I pushed the fix almost two weeks ago. So let me know your thoughts.

pulled your changes, checks are running [here](https://github.com/rhinofi/android-nixpkgs-test/actions)

adrian-gierakowski commented:
All looks good. Note that the ci jobs only check the architecture of the emulator. I can test them with a real project on Monday if you’d like to wait for that

hadilq commented:
> All looks good. Note that the ci jobs only check the architecture of the emulator. I can test them with a real project on Monday if you’d like to wait for that 

Sounds good. Thank you!

adrian-gierakowski commented:
> > All looks good. Note that the ci jobs only check the architecture of the emulator. I can test them with a real project on Monday if you’d like to wait for that
> 
> Sounds good. Thank you!

Apologies for the delay, looking into it now

adrian-gierakowski commented:
I was able to run the emulator in headless mode both on x86_64-linux and aarch64-darwin (when built as can be seen [here](https://github.com/rhinofi/android-nixpkgs-test/blob/ef031810d7d2bb837dd58e09f36dbb645ed38bab/flake.nix#L30-L46)), but I'm having some issues with dev shell  created via [devenv's android module](https://github.com/cachix/devenv/blob/991abff153b995192bf36655394246fc97ba8627/src/modules/integrations/android.nix) which seems to include the x86 emulator on aarch64-darwin. Don't have the time to debug it further now, but can get back to it early next week

hadilq commented:
> I was able to run the emulator in headless mode both on x86_64-linux and aarch64-darwin (when built as can be seen [here](https://github.com/rhinofi/android-nixpkgs-test/blob/ef031810d7d2bb837dd58e09f36dbb645ed38bab/flake.nix#L30-L46)), but I'm having some issues with dev shell created via [devenv's android module](https://github.com/cachix/devenv/blob/991abff153b995192bf36655394246fc97ba8627/src/modules/integrations/android.nix) which seems to include the x86 emulator on aarch64-darwin. Don't have the time to debug it further now, but can get back to it early next week

No worries. Thank you for checking it!

the cachix/devenv is new to me, so I couldn't find out how to set the `inputs.nixpkgs` to the HEAD commit of this PR! Also I found the [failed test](https://github.com/cachix/devenv/actions/runs/12462416205/job/34785238539#step:13:240) in that repository, where the main problem looks like this
```
[2024-12-23 07:44:59][Info][error:FailureResponse (Request {requestPath = (BaseUrl {baseUrlScheme = Https, baseUrlHost = "cachix.org", baseUrlPort = 443, baseUrlPath = ""},"/api/v1/cache/devenv/multipart-nar"), requestQueryString = fromList [("compression",Just "zst")], requestBody = Nothing, requestAccept = fromList [application/json;charset=utf-8,application/json], requestHeaders = fromList [("Authorization","<REDACTED>")], requestHttpVersion = HTTP/1.1, requestMethod = "POST"}) (Response {responseStatusCode = Status {statusCode = 502, statusMessage = "Bad Gateway"}, responseHeaders = fromList [("Server","cloudflare"),("Date","Mon, 23 Dec 2024 07:44:59 GMT"),("Content-Type","text/html"),("Content-Length","155"),("Connection","keep-alive"),("CF-RAY","8f66bac168f78d79-HEL")], responseHttpVersion = HTTP/1.1, responseBody = "<html>\r\n<head><title>502 Bad Gateway</title></head>\r\n<body>\r\n<center><h1>502 Bad Gateway</h1></center>\r\n<hr><center>cloudflare</center>\r\n</body>\r\n</html>\r\n"})] Failed /nix/store/3yk79xqb46lgzm8wdmwabdhcxcwxlb4j-devenv-test
```
Not sure what I am looking at, or even this is what you have referring to!

adrian-gierakowski commented:
@hadilq you can override pkgs specialArg (used [here](https://github.com/cachix/devenv/blob/991abff153b995192bf36655394246fc97ba8627/src/modules/integrations/android.nix#L6)) by following the instructions on how to use [devenv with flakes](https://devenv.sh/guides/using-with-flakes/) and setting pkgs argument in call to mkShell to one based on this PR

hadilq commented:
Hey @adrian-gierakowski I just found some time to take a further look at this problem. I see that all the Android tests in devenv are passing now! It's not clear why we blocked this PR! BTW, isn't devenv's problem out of scope for this PR? I think we should merge this one, since it already solved the problem in https://github.com/NixOS/nixpkgs/issues/359918, then later fix up comming issues. What do you think?

hadilq commented:
Let me rebase and squish it, to be prepared for the approval.

adrian-gierakowski commented:
> Hey @adrian-gierakowski I just found some time to take a further look at this problem. I see that all the Android tests in devenv are passing now! It's not clear why we blocked this PR! BTW, isn't devenv's problem out of scope for this PR? I think we should merge this one, since it already solved the problem in #359918, then later fix up comming issues. What do you think?

Yes, it was most likely a devenv issue. My intention was to test this PR with a project which both builds and then runs an app in an emulator. Let me try again tonight.

hadilq commented:
Hello @adrian-gierakowski, let's follow up with this PR again! :D Let me know what do you think!

numinit commented:
Ack, somehow missed this one. I think we can ping the Android team in the future. I'll give this a test.

numinit commented:
Shell works fine.

numinit commented:
Is a mirror unstable?

hadilq commented:
> Is a mirror unstable?
> 
> Edit: Before you commit, make sure to run generate.sh one more time. (It worked with API 35 after that)

Yes I noticed this too! The Google's upstream packages are not stable and sometime they are changing, so we have to run `generate.sh` again!

hadilq commented:
I updated this branch with master. I think it's ready to merge.

RossComputerGuy reviewed (commented):
This looks cool, I've heard from @numinit about aarch64 host support. Does this contain the necessary changes? I might give it a try on my Ampere system .

hadilq commented:
> This looks cool, I've heard from @numinit about aarch64 host support. Does this contain the necessary changes? I might give it a try on my Ampere system .

Yep! It's cool! :D It would be great if we have more tests, so please check it out! Also Adrian had some [tests in another repo](https://github.com/rhinofi/android-nixpkgs-test/actions) where I think it would be great to have them here. Do you know how to make a matrix test here in nixpkgs repo?

hadilq commented:
@ofborg build -A androidenv.test-suite.tests  x86_64-linux x86_64-darwin aarch64-darwin

hadilq commented:
@ofborg build androidenv.test-suite.tests

hadilq commented:
> > Do you know how to make a matrix test here in nixpkgs repo?
> 
> Yes, you can `@ofborg build a b c` where a b c different attribute names in `pkgs`.

@RossComputerGuy, it [skipped](https://github.com/NixOS/nixpkgs/pull/360151/checks?check_run_id=37107951975)! What am I missing?


```
OfBorg / androidenv.test-suite.tests on aarch64-darwin

No attempt

The following builds were skipped because they don't evaluate on aarch64-darwin: androidenv.test-suite.tests
Details
Partial log

Cannot nix-instantiate `androidenv.test-suite.tests` because:
error: access to absolute path '/private' is forbidden in restricted mode

```

RossComputerGuy commented:
I'm not sure, ask the folks in infra or the Ofborg matrix.

hadilq commented:
> I'm not sure, ask the folks in infra or the Ofborg matrix.

Thank you! Let's see if I can run those tests in this PR :D

RossComputerGuy commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 360151`

RossComputerGuy commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 360151`

---
### `aarch64-linux`

RossComputerGuy commented:
![image](https://github.com/user-attachments/assets/28b7c2bb-5935-4131-a009-fa6c4fee0fef)
Doesn't eval

RossComputerGuy commented:
![Screenshot_20250212-140102](https://github.com/user-attachments/assets/7ab67deb-85c3-48e8-a6b7-866eb2f3fa6c)
Doesn't eval

numinit commented:
I'll give it a check on the community builders.

numinit commented:
Alright, so... yeah. Bad news.

I think there is a more limited set of attributes that can eval on aarch64, because Google's prebuilts support aarch64-darwin but not aarch64-linux. Apply the following patch and it will become more obvious.

```patch
From 6d329230b378985f2d628df534b352c6885b2de3 Mon Sep 17 00:00:00 2001
From: Morgan Jones <me@numin.it>
Date: Fri, 14 Feb 2025 05:23:33 +0000
Subject: [PATCH] androidenv: make i686 packages only eval on x86_64

---
 .../mobile/androidenv/build-tools.nix         |  5 ++--
 .../androidenv/compose-android-packages.nix   | 24 +++++++++++--------
 .../mobile/androidenv/emulator.nix            |  5 ++--
 .../mobile/androidenv/test-suite.nix          |  2 +-
 pkgs/development/mobile/androidenv/tools.nix  |  4 ++--
 5 files changed, 22 insertions(+), 18 deletions(-)

diff --git a/pkgs/development/mobile/androidenv/build-tools.nix b/pkgs/development/mobile/androidenv/build-tools.nix
index 5e60dbf3234e..53a80dd14967 100644
--- a/pkgs/development/mobile/androidenv/build-tools.nix
+++ b/pkgs/development/mobile/androidenv/build-tools.nix
@@ -1,10 +1,11 @@
-{deployAndroidPackage, lib, package, os, autoPatchelfHook, makeWrapper, pkgs, pkgsi686Linux, postInstall}:
+{deployAndroidPackage, lib, stdenv, package, os, autoPatchelfHook, makeWrapper, pkgs, pkgsi686Linux, postInstall}:
 
 deployAndroidPackage {
   inherit package;
   nativeBuildInputs = [ makeWrapper ]
     ++ lib.optionals (os == "linux") [ autoPatchelfHook ];
-  buildInputs = lib.optionals (os == "linux") [ pkgs.glibc pkgs.zlib pkgs.ncurses5 pkgsi686Linux.glibc pkgsi686Linux.zlib pkgsi686Linux.ncurses5 pkgs.libcxx ];
+  buildInputs = lib.optionals (os == "linux") [ pkgs.glibc pkgs.zlib pkgs.ncurses5 pkgs.libcxx ]
+    ++ lib.optionals (os == "linux" && stdenv.isx86_64) (with pkgsi686Linux; [ glibc zlib ncurses5 ]);
   patchInstructions = ''
     ${lib.optionalString (os == "linux") ''
       addAutoPatchelfSearchPath $packageBaseDir/lib
diff --git a/pkgs/development/mobile/androidenv/compose-android-packages.nix b/pkgs/development/mobile/androidenv/compose-android-packages.nix
index ff9897b8dbe0..61317e13950a 100644
--- a/pkgs/development/mobile/androidenv/compose-android-packages.nix
+++ b/pkgs/development/mobile/androidenv/compose-android-packages.nix
@@ -74,17 +74,21 @@ let
     lib.attrsets.mapAttrsRecursive
       (path: value:
         if (builtins.elemAt path ((builtins.length path) - 1)) == "archives" then
-          lib.last (builtins.map (archive:
+          let
+            validArchives = builtins.filter (archive:
+              let
+                isTargetOs = if builtins.hasAttr "os" archive then
+                  archive.os == os || archive.os == "all" else true;
+                isTargetArc = if builtins.hasAttr "arch" archive then
+                  archive.arch == arch || archive.arch == "all" else true;
+              in
+                isTargetOs && isTargetArc
+            ) value;
+          in
+          assert lib.assertMsg (builtins.length validArchives > 0) "No valid archives for ${lib.concatMapStringsSep "." (x: ''"${x}"'') path} for os=${os}, arch=${arch}";
+          lib.last (map (archive:
             (fetchurl { inherit (archive) url sha1; })
-          ) (builtins.filter (archive:
-            let
-              isTargetOs = if builtins.hasAttr "os" archive then
-                archive.os == os || archive.os == "all" else true;
-              isTargetArc = if builtins.hasAttr "arch" archive then
-                archive.arch == arch || archive.arch == "all" else true;
-            in
-              isTargetOs && isTargetArc
-          ) value))
+          ) validArchives)
         else value
       )
       attrSet;
diff --git a/pkgs/development/mobile/androidenv/emulator.nix b/pkgs/development/mobile/androidenv/emulator.nix
index 47c84692e688..765bf40df364 100644
--- a/pkgs/development/mobile/androidenv/emulator.nix
+++ b/pkgs/development/mobile/androidenv/emulator.nix
@@ -1,4 +1,4 @@
-{ deployAndroidPackage, lib, package, os, autoPatchelfHook, makeWrapper, pkgs, pkgsi686Linux, postInstall }:
+{ deployAndroidPackage, lib, stdenv, package, os, autoPatchelfHook, makeWrapper, pkgs, pkgsi686Linux, postInstall }:
 
 deployAndroidPackage {
   inherit package;
@@ -16,7 +16,6 @@ deployAndroidPackage {
       ncurses5
       libdrm
       stdenv.cc.cc
-      pkgsi686Linux.glibc
       expat
       freetype
       nss
@@ -36,7 +35,7 @@ deployAndroidPackage {
       libICE
       libSM
       libxkbfile
-    ]);
+    ]) ++ lib.optional (os == "linux" && stdenv.isx86_64) pkgsi686Linux.glibc;
   patchInstructions = lib.optionalString (os == "linux") ''
     addAutoPatchelfSearchPath $packageBaseDir/lib
     addAutoPatchelfSearchPath $packageBaseDir/lib64
diff --git a/pkgs/development/mobile/androidenv/test-suite.nix b/pkgs/development/mobile/androidenv/test-suite.nix
index 0b6984366191..1db990e416b0 100644
--- a/pkgs/development/mobile/androidenv/test-suite.nix
+++ b/pkgs/development/mobile/androidenv/test-suite.nix
@@ -13,7 +13,7 @@ let
 in
 stdenv.mkDerivation {
   name = "androidenv-test-suite";
-  buidInputs = lib.mapAttrsToList (name: value: value) all-tests;
+  buildInputs = lib.mapAttrsToList (name: value: value) all-tests;
 
   buildCommand = ''
     touch $out
diff --git a/pkgs/development/mobile/androidenv/tools.nix b/pkgs/development/mobile/androidenv/tools.nix
index 1938ff4cca60..6cd41305bebb 100644
--- a/pkgs/development/mobile/androidenv/tools.nix
+++ b/pkgs/development/mobile/androidenv/tools.nix
@@ -1,4 +1,4 @@
-{deployAndroidPackage, lib, package, autoPatchelfHook, makeWrapper, os, pkgs, pkgsi686Linux, postInstall}:
+{deployAndroidPackage, lib, stdenv, package, autoPatchelfHook, makeWrapper, os, pkgs, pkgsi686Linux, postInstall}:
 
 deployAndroidPackage {
   name = "androidsdk-tools";
@@ -10,7 +10,7 @@ deployAndroidPackage {
         stdenv.cc.cc.libgcc or null # fix for https://github.com/NixOS/nixpkgs/issues/226357
       ])
       ++ (with pkgs.xorg; [ libX11 libXrender libXext ])
-      ++ (with pkgsi686Linux; [ glibc xorg.libX11 xorg.libXrender xorg.libXext fontconfig.lib freetype zlib ])
+      ++ lib.optionals (os == "linux" && stdenv.isx86_64) (with pkgsi686Linux; [ glibc xorg.libX11 xorg.libXrender xorg.libXext fontconfig.lib freetype zlib ])
     );
 
   patchInstructions = ''
-- 
2.47.0
```

```
$ NIXPKGS_ACCEPT_ANDROID_SDK_LICENSE=1 NIXPKGS_ALLOW_UNFREE=1 nix-build -A androidenv.test-suite
       error: No valid archives for "cmake"."3.6.4111459"."archives" for os=linux, arch=aarch64
```

Looking in the repo.json, there are no prebuilds for the NDK, cmake, or emulator for aarch64. Long term, we should unvendor and build these, since they are mostly buildable: see https://github.com/lzhiyong/termux-ndk/releases. For now, we should just error out but have a meaningful error if people use incompatible packages.

numinit commented:
@hadilq Do you want to apply that patch in your tree? I think that will provide more meaningful error messages for people who try to, for instance, use the NDK, but they can still use the SDKs.

hadilq commented:
> @hadilq Do you want to apply that patch in your tree? I think that will provide more meaningful error messages for people who try to, for instance, use the NDK, but they can still use the SDKs.

@numinit Thank you Morgan! It's exactly what we needed here.

@RossComputerGuy would you check it out again? Thanks!

numinit commented:
I think we should make one more change so androidenv.androidPkgs works:

https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/mobile/androidenv/default.nix#L32

Either those includeNdk/includeEmulator/includeSystemImages defaults should all be `system == "x86_64-linux" || system == "aarch64-darwin" or we need to add an "if-supported" default there which is a little more aware of how the repo works.

hadilq commented:
> I think we should make one more change so androidenv.androidPkgs works:
> 
> https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/mobile/androidenv/default.nix#L32
> 
> Either those includeNdk/includeEmulator/includeSystemImages defaults should all be `system == "x86_64-linux" || system == "aarch64-darwin"` or we need to add an "if-supported" default there which is a little more aware of how the repo works. Maybe I'll try the latter approach in the next day or two.
> 
> Edit: We should also check it on the Darwin community builder and fix the tests.

I pushed https://github.com/NixOS/nixpkgs/pull/360151/commits/1bdb5e036a23f78f77d9d29913c450d9fcad8fd5, so let me know if I addressed these concerns correctly!

hadilq commented:
@ofborg build androidenv.test-suite.tests

numinit commented:
Not 100% what I was going for: people should be able to build it on aarch64-linux (and the androidenv.androidPkgs attribute should work) - they just won't get the NDK or emulator.

hadilq commented:
@numinit Please check [d370abe](https://github.com/NixOS/nixpkgs/pull/360151/commits/d370abe709fc56fb0b45b4a5ce8cc16b41243418)

numinit commented on pkgs/development/mobile/androidenv/compose-android-packages.nix:
```diff
@@ -71,9 +97,21 @@ let
+            ) value;
+          in
+          assert lib.assertMsg (builtins.length validArchives > 0) "No valid archives for ${lib.concatMapStringsSep "." (x: ''"${x}"'') path} for os=${os}, arch=${arch}";
+          lib.last (map (archive:
```
Would it be easier to replace this with null if there's no supported archive, instead of doing this assert check? That way we don't need any external logic. If there's no supported binary for the platform we can use that as the predicate.

hadilq replied:
Sorry, but in my mind it's clear to have an assertion over a null. However, feel free to send the patch and I'll apply it.

numinit replied:
Fair. I wonder if we'd have to tryEval or something.

In any case, I think if we'd like to submit it this way after @rosscomputerguy checks on aarch64 that's fine with me as it'll unblock people.

numinit commented on pkgs/development/mobile/androidenv/default.nix:
```diff
@@ -29,9 +29,15 @@ rec {
+      system == "x86_64-linux" || system == "x86_64-darwin" || system == "aarch64-darwin";
+    includeNDK =
+      with pkgs.stdenv.hostPlatform;
+      system == "x86_64-linux" || system == "x86_64-darwin" || system == "aarch64-darwin";
```
Maybe we should pass an "if-supported" string here. Then if it's "if-supported" we can just not include it instead of erroring.

hadilq replied:
I think I have problem understanding what would be the implementation for "if-supported"! would you explain more? Thanks!

numinit replied:
If there's a NDK package or similar for the current arch, we link it in. Otherwise we just eval warn about it.

numinit commented on pkgs/development/mobile/androidenv/compose-android-packages.nix:
```diff
@@ -26,10 +26,36 @@
+  includeSystemImagesInternal = lib.warnIf (!supportedBinariesInternal && includeSystemImages)
+    "SystemImages are not supported for ${stdenv.hostPlatform.system}"
+    (supportedBinariesInternal && includeSystemImages);
+  includeNDKInternal =  lib.warnIf (!supportedBinariesInternal && includeNDK)
```
All this logic will be way simpler if we just check for whether the archive exists after it's been filtered for the correct arch and OS 😄

hadilq replied:
my problem with that approach is all the "all" tags! if we filter for a specific arch-os, and it doesn't exist, then should be take the "all" or should we go with null?! I probably need more time to play around with these!!

numinit replied:
Yeah, I will need to play around with it over the next day or two, I think

hadilq replied:
I pushed again! Please check [1c0661a](https://github.com/NixOS/nixpkgs/pull/360151/commits/1c0661a0f1f11e778308afa1e047b97d75affc71)

numinit replied:
Thanks, will do Hadi!

hadilq replied:
I just got access to the community-builders and start to getting more insight into the problem! so above commit is not the last one! :D

JohnRTitor commented on pkgs/development/mobile/androidenv/tools.nix:
```diff
@@ -1,16 +1,16 @@
       ])
       ++ (with pkgs.xorg; [ libX11 libXrender libXext ])
-      ++ (with pkgsi686Linux; [ glibc xorg.libX11 xorg.libXrender xorg.libXext fontconfig.lib freetype zlib ])
+      ++ lib.optionals (os == "linux" && stdenv.isx86_64) (with pkgsi686Linux; [ glibc xorg.libX11 xorg.libXrender xorg.libXext fontconfig.lib freetype zlib ])
```
aah, can we not use `stdenv.hostPlatform.system == "x86_64-linux"`

hadilq replied:
I applied this, however, I noticed it's a longer string in the end, so what's the point! I reverted!

numinit commented:
Evals locally on x86_64 and on the aarch64-{linux,darwin} community builders while running:

`NIXPKGS_ACCEPT_ANDROID_SDK_LICENSE=1 NIXPKGS_ALLOW_UNFREE=1 nix-build -A androidenv.androidPkgs.androidsdk`

And things even work on both x86_64-linux and aarch64-darwin!

Though this happens on aarch64-linux:

```
./result/libexec/android-sdk/build-tools/35.0.0/aapt2
-bash: ./result/libexec/android-sdk/build-tools/35.0.0/aapt2: cannot execute binary file: Exec format error
```

Confused, I looked at the repo.json and discovered that the repos misrepresent things like the build-tools as arch "all" and this happens with cmake too but _not_ the NDK :man_facepalming:. No wonder this problem was confusing. It'd probably work with binfmt_misc emulating an x64 machine though.

![Screenshot_20250215_162507](https://github.com/user-attachments/assets/2edb612c-9428-4c75-9bdf-c3ce3e591ae2)

----

And indeed I get warnings on aarch64-linux and Darwin while running:

`NIXPKGS_ACCEPT_ANDROID_SDK_LICENSE=1 NIXPKGS_ALLOW_UNFREE=1 nix-build -A androidenv.test-suite`

Though the test suite does fail to run there and passes locally, but is also looking for arch-specific packages. The error was super weird on aarch64-linux though, it failed to build in general:

```
error: builder for '/nix/store/c7h9bb9gb9al9s01pjyzdmq46jrpv4k2-android-sdk-cmake-3.6.4111459.drv' failed with exit code 1;
       last 9 log lines:
       > Running phase: unpackPhase
       > Running phase: patchPhase
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > no configure script, doing nothing
       > Running phase: buildPhase
       > no Makefile or custom buildPhase, doing nothing
       > Running phase: installPhase
       > /nix/store/apz6fmxvk7h8fd5qm4nlnf5xjd19n0zg-stdenv-linux/setup: line 1767: cd: /build/extractedzip-0: No such file or directory
```

This is likely best explained with cmake and the other tools being marked as platform "all" when that is simply not true.

----

I think the combination of this and #379534 will fix the issues we have with updates, I'll rebase it right after this lands. Maybe I'll look into the test issues too, though I'm good with the approach we've come up with in the meantime. In the long term I'll probably do a nixfmt and best-practices pass on all this code since there are certainly some things I'd do differently compared to how androidenv was originally written. But man the Google repos are not helping the cross-arch compatibility here :-(

numinit commented:
Oh dear it's worse than I thought. Here's an example where host-arch is completely missing for one copy of the NDK.

```xml
  <remotePackage path="ndk;16.1.4479499">
    <type-details xsi:type="generic:genericDetailsType"/>
    <revision>
      <major>16</major>
      <minor>1</minor>
      <micro>4479499</micro>
    </revision>
    <display-name>NDK (Side by side) 16.1.4479499</display-name>
    <uses-license ref="android-sdk-license"/>
    <channelRef ref="channel-0"/>
    <archives>
      <archive>
        <complete>
          <size>839630771</size>
          <checksum type="sha1">e51e615449b98c716cf912057e2682e75d55e2de</checksum>
          <url>android-ndk-r16b-darwin-aarch64.zip</url>
        </complete>
        <host-os>macosx</host-os>
        <host-arch>aarch64</host-arch>
      </archive>
      <archive>
        <complete>
          <size>839630771</size>
          <checksum type="sha1">e51e615449b98c716cf912057e2682e75d55e2de</checksum>
          <url>android-ndk-r16b-darwin-x86_64.zip</url>
        </complete>
        <host-os>macosx</host-os>
        <host-arch>x64</host-arch>
      </archive>
      <archive>
        <complete>
          <size>852525873</size>
          <checksum type="sha1">42aa43aae89a50d1c66c3f9fdecd676936da6128</checksum>
          <url>android-ndk-r16b-linux-x86_64.zip</url>
        </complete>
        <host-os>linux</host-os>
        <host-arch>x64</host-arch>
      </archive>
      <archive>
        <complete>
          <size>656720029</size>
          <checksum type="sha1">becaf3d445a4877ca1a9300a62f0934a4838c7fa</checksum>
          <url>android-ndk-r16b-windows-x86.zip</url>
        </complete>
        <host-os>windows</host-os>
        <host-arch>x86</host-arch>
      </archive>
      <archive>
        <complete>
          <size>723301086</size>
          <checksum type="sha1">f3f1909ed1052e98dda2c79d11c22f3da28daf25</checksum>
          <url>android-ndk-r16b-windows-x86_64.zip</url>
        </complete>
        <host-os>windows</host-os>
        <host-arch>x64</host-arch>
      </archive>
    </archives>
  </remotePackage>
  <remotePackage path="ndk;26.2.11394342">
    <type-details xsi:type="generic:genericDetailsType"/>
    <revision>
      <major>26</major>
      <minor>2</minor>
      <micro>11394342</micro>
    </revision>
    <display-name>NDK (Side by side) 26.2.11394342</display-name>
    <uses-license ref="android-sdk-license"/>
    <channelRef ref="channel-0"/>
    <archives>
      <archive>
        <complete>
          <size>983592055</size>
          <checksum type="sha1">ceb609baf514614015279b5eee7dffc859426952</checksum>
          <url>android-ndk-r26c-darwin.zip</url>
        </complete>
        <host-os>macosx</host-os>
      </archive>
      <archive>
        <complete>
          <size>668556021</size>
          <checksum type="sha1">7faebe2ebd3590518f326c82992603170f07c96e</checksum>
          <url>android-ndk-r26c-linux.zip</url>
        </complete>
        <host-os>linux</host-os>
      </archive>
      <archive>
        <complete>
          <size>660158794</size>
          <checksum type="sha1">f8c8aa6135241954461b2e3629cada4722e13ee7</checksum>
          <url>android-ndk-r26c-windows.zip</url>
        </complete>
        <host-os>windows</host-os>
      </archive>
    </archives>
  </remotePackage>
```

numinit commented:
Let me try to fix that. I think it may be possible with mkrepo.rb to default the archs to something reasonable if they're missing based on the host OS, since there are no aarch64-linux or x86_64-darwin for the newer packages, and the older ones explicitly have it provided.

numinit commented:
One hangup I'm running into is that we need to do this for historical packages, too. I looked at fixing it in Nix and think we ought to just fix the data Nix is reading from (aka the repo JSON). Annoying.

numinit commented:
Ahhhhhh it's even worse. Old packages are sometimes truly ambiguous about whether they're for x86-64 Darwin or aarch64 Darwin. I guess on Darwin we could do "all" for the default arch due to Rosetta. On Linux they've never shipped aarch64 binaries so it's at least less ambiguous.

numinit commented:
Seems like the repositories are just broken like this:

https://issuetracker.google.com/issues/387312226

numinit commented:
Correction:

```
numinit@darwin01:~/nixpkgs/ > file ./result/bin/adb
./result/bin/adb: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit executable x86_64] [arm64]
./result/bin/adb (for architecture x86_64):     Mach-O 64-bit executable x86_64
./result/bin/adb (for architecture arm64):      Mach-O 64-bit executable arm64
```

So Darwin should just support both since they use fat binaries.

numinit reviewed (approved):
I think we have to do a little more work to make the architectures more specific than 'all', but it's working on x86_64-linux and aarch64-darwin. Google's repos are just broken and probably we'll have to download the files as we're updating them and guess the architecture in order to fix that.

hadilq commented:
The last push, [90c8b98](https://github.com/NixOS/nixpkgs/pull/360151/commits/90c8b98330ccd14e39434f9959ac2f38d542094c), fixes the problems while running
```
NIXPKGS_ALLOW_UNFREE=1 nix-build -A androidenv.test-suite
```
on `aarch64-darwin` host, and provides enough warnings for `aarch64-linux` host. Let me know if you have more concerns.

cc: @numinit @RossComputerGuy

RossComputerGuy commented:
```
evaluation warning: No valid archives for "cmake"."3.6.4111459"."archives" for os=linux, arch=aarch64
evaluation warning: No valid archives for "extras;google;auto"."archives" for os=linux, arch=aarch64
evaluation warning: No valid archives for "emulator"."35.5.2"."archives" for os=linux, arch=aarch64
evaluation warning: No valid archives for "emulator"."35.5.2"."archives" for os=linux, arch=aarch64
```
Nice

numinit commented:
Awesome.

numinit reviewed (approved):
Works on aarch64-darwin and x86_64-linux.

numinit commented:
I think I'm going to submit a Google bugtracker ticket tomorrow. It's a low chance they fix it by making the arch of their packages correct in the repository XML, but that's the root cause.

numinit commented:
I think this is good to merge. #379534 will make the updates and related test runs automatic.

JohnRTitor commented on pkgs/development/mobile/androidenv/examples/shell.nix:
```diff
@@ -16,7 +16,10 @@
+  config ? pkgs.config,
+  # You probably need to set it to true to express consent.
+  licenseAccepted ?
+    config.android_sdk.accept_license or (builtins.getEnv "NIXPKGS_ACCEPT_ANDROID_SDK_LICENSE" == "1"),
```
Humm, as I understand it, `or` on Nix isn't actually `||`, but rather a fallback operator. If `config.android_sdk.accept_license` is undefined, it'll fallback to checking the env (with `--impure`).

But what about this scenario? Say: you have it disabled in your nixpkgs config, but still want to temporarily build this. It will still be unable to build in that case.

hadilq replied:
Did you test your scenario? I just tested it by setting accept_license to false, but using env variable to accept them and the build didn't complain! Let me know if I didn't apply the scenario correctly!

numinit replied:
It works due to the two levels of defaulting:

```
nix-repl> let config.android_sdk.accept_license = false; f = { licenseAccepted ? config.android_sdk.accept_license or (builtins.getEnv "NIXPKGS_ACCEPT_ANDROID_SDK_LICENSE" == "1") }: licenseAccepted; in f {}
false

nix-repl> let config.android_sdk.accept_license = false; f = { licenseAccepted ? config.android_sdk.accept_license or (builtins.getEnv "NIXPKGS_ACCEPT_ANDROID_SDK_LICENSE" == "1") }: licenseAccepted; in f { licenseAccepted = true; }
true

nix-repl> let config.android_sdk.accept_license = true; f = { licenseAccepted ? config.android_sdk.accept_license or (builtins.getEnv "NIXPKGS_ACCEPT_ANDROID_SDK_LICENSE" == "1") }: licenseAccepted; in f { }
true
```

numinit replied:
@hadilq I think we pass down the top level env into all the other derivations right?

numinit replied:
OK, yeah, I can actually repro that I think.

```
$ export NIXPKGS_ACCEPT_ANDROID_SDK_LICENSE=1
$ nix repl 
Nix 2.24.12
Type :? for help.
nix-repl> builtins.getEnv "NIXPKGS_ACCEPT_ANDROID_SDK_LICENSE" 
"1"

nix-repl> with import ./. { config = { allowUnfree = true; android_sdk.accept_license = false; }; }; pkgs.androidenv.androidPkgs.androidsdk      
       error: --- android-sdk-license ---
       Terms and Conditions

       This is the Android Software Development Kit License Agreement

```

numinit commented:
#379534 good to go next. :-)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/androidenv-soon-to-be-automatically-tracking-latest-in-a-nixpkgs-near-you/60639/1
